src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -111,8 +111,8 @@ public UnfilteredPartitionIterator search(ReadExecutionController executionContr
         if (!command.isTopK())
             return queryIndexes.get();
 
-        // VSTODO performance: if there is shadowed primary keys, we have to at least query twice.
-        //  First time to find out there are shawdow keys, second time to find out there are no more shadow keys.
+        // If there are shadowed primary keys, we have to at least query twice.
+        // First time to find out there are shadowed keys, second time to find out there are no more shadow keys.
```

Comment:
Only asking because you removed the TODO. We keep track of the shadowed primary keys and exclude them from the graph search. Is it worth doing the same for the PKs that we know are _not_ shadowed so we could decrease the number of vectors we're searching for?

Reply:
I removed the TODO b/c it sounded like making multiple passes was something we can/should fix which I don't think is true

Yours is an interesting idea that is worth following up on another time :)

Reply:
Definitely another time, let's get this merged :)